cmd/agent/common/autodiscovery.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+func setupAutoDiscovery(confSearchPaths []string, metaScheduler *scheduler.MetaScheduler) *autodiscovery.AutoConfig {
+	ad := autodiscovery.NewAutoConfig(metaScheduler)
+	ad.AddConfigProvider(providers.NewFileConfigProvider(confSearchPaths), false, 0)
+
+	// Autodiscovery cannot easily use config.RegisterOverrideFunc() due to Unmarshalling
+	var discoveredProviders []config.ConfigurationProviders
+	var discoveredListeners []config.Listeners
+	if config.Datadog.GetBool("autoconf_from_environment") {
+		discoveredProviders, discoveredListeners = discoverAutodiscoveryComponents()
+	}
+
+	// Register additional configuration providers
+	var CP []config.ConfigurationProviders
+	err := config.Datadog.UnmarshalKey("config_providers", &CP)
+	if err == nil {
+		// Add extra config providers
+		for _, name := range config.Datadog.GetStringSlice("extra_config_providers") {
+			CP = append(CP, config.ConfigurationProviders{Name: name, Polling: true})
+		}
+
+		for _, provider := range discoveredProviders {
+			alreadyPresent := false
+			for _, existingProvider := range CP {
+				if existingProvider.Name == provider.Name {
+					alreadyPresent = true
+					break
+				}
+			}
+
+			if !alreadyPresent {
+				CP = append(CP, provider)
+			}
+		}
+	} else {
+		log.Errorf("Error while reading 'config_providers' settings: %v", err)
+	}
+
+	// Adding all found providers
+	for _, cp := range CP {
+		factory, found := providers.ProviderCatalog[cp.Name]
+		if found {
+			configProvider, err := factory(cp)
+			if err == nil {
+				pollInterval := providers.GetPollInterval(cp)
+				if cp.Polling {
+					log.Infof("Registering %s config provider polled every %s", cp.Name, pollInterval.String())
+				} else {
+					log.Infof("Registering %s config provider", cp.Name)
+				}
+				ad.AddConfigProvider(configProvider, cp.Polling, pollInterval)
+			} else {
+				log.Errorf("Error while adding config provider %v: %v", cp.Name, err)
+			}
+		} else {
+			log.Errorf("Unable to find this provider in the catalog: %v", cp.Name)
+		}
+	}
+
+	var listeners []config.Listeners
+	err = config.Datadog.UnmarshalKey("listeners", &listeners)
+	if err == nil {
+		// Add extra listeners
+		for _, name := range config.Datadog.GetStringSlice("extra_listeners") {
+			listeners = append(listeners, config.Listeners{Name: name})
+		}
+
+		for _, listener := range discoveredListeners {
+			alreadyPresent := false
+			for _, existingListener := range listeners {
+				if listener.Name == existingListener.Name {
+					alreadyPresent = true
+					break
+				}
+			}
+
+			if !alreadyPresent {
+				listeners = append(listeners, listener)
+			}
+		}
+
+		ad.AddListeners(listeners)
+	} else {
+		log.Errorf("Error while reading 'listeners' settings: %v", err)
+	}
+
+	return ad
+}
+
+func discoverAutodiscoveryComponents() ([]config.ConfigurationProviders, []config.Listeners) {
```

Comment:
```suggestion
func discoverAutodiscoveryComponents() ([]config.ConfigurationProviders, config.Listeners) {
```
we must not enable more than one service listener here, if docker and kubelet both available, only the kubelet listener must be enabled

Reply:
I think signature should still be `[]config.Listeners` as it's not limited to this use case, but I'll add the logic with Docker/Kubelet